File: datasets/CRUST_bench/CBench/Genetic-neural-network-for-simple-control/test/tests/general/test_system_builder.c
#include "general/test_system_builder.h"
#include "general/systems_builder.h"

#include "general/pid_controller.h"

#include <stdlib.h>
#include <stdio.h>

// used to print testing outputs
#define ANSI_COLOR_RED     "\x1b[31m"
#define ANSI_COLOR_GREEN   "\x1b[32m"
#define ANSI_BOLD         "\x1b[1m"
#define ANSI_COLOR_RESET   "\x1b[0m"

int testSelectSystem(){
  printf(ANSI_BOLD "=======TEST SELECT SYSTEM STARTED=======" ANSI_COLOR_RESET "\n");

  
  float (*func_ptr)(float);
  struct PID *pid = (struct PID*)malloc(sizeof(struct PID));
  selectSystem(&func_ptr);

  pid->signal = (struct Signal*)malloc(sizeof(struct Signal));
  pid->output = (struct Signal*)malloc(sizeof(struct Signal));

  pid->signal->signal = (float*)malloc(sizeof(float));
  pid->output->signal = (float*)malloc(sizeof(float));

  pid->dataSystem = (float*)malloc(sizeof(float));

  deletePid(pid);

  printf(ANSI_BOLD ANSI_COLOR_GREEN "=======TEST SELECT SYSTEM SUCCESSFUL=======" ANSI_COLOR_RESET "\n");
  return 1;
}

File: datasets/CRUST_bench/RBench/Genetic_neural_network_for_simple_control/src/bin/test_system_builder.rs
use Genetic_neural_network_for_simple_control::systems_builder::selectSystem;
use Genetic_neural_network_for_simple_control::pid_controller::{deletePid, PID};
use Genetic_neural_network_for_simple_control::signal_designer::Signal;
#[test]
pub fn testSelectSystem() {
    println!("\x1b[1m=======TEST SELECT SYSTEM STARTED=======\x1b[0m");

    let mut func_ptr: Option<fn(&mut [f32]) -> f32> = None;
    let mut pid = Box::new(PID::default());

    // Select the system
    let _size = selectSystem(&mut func_ptr);

    // Allocate dummy signals and system memory
    pid.signal = Some(Box::new(Signal {
        signal: vec![0.0],
        dt: 0.1,
        length: 1,
    }));

    pid.output = Some(Box::new(Signal {
        signal: vec![0.0],
        dt: 0.1,
        length: 1,
    }));

    pid.dataSystem = Some(Box::new([0.0; 10]));

    deletePid(pid.as_mut());

    println!("\x1b[1m\x1b[32m=======TEST SELECT SYSTEM SUCCESSFUL=======\x1b[0m");
    
}
fn main(){}